ARROW-3966 [Java] JDBC Column Metadata in Arrow Field Metadata

mikepigott opened the pull request:
https://issues.apache.org/jira/browse/ARROW-3966

This change includes #3133, and supports a new configuration item called "Include Metadata."  If true, metadata from the JDBC ResultSetMetaData object is pulled along to the Schema Field Metadata.  For now, this includes:
* Catalog Name
* Table Name
* Column Name
* Column Type Name

mikepigott commented:
https://travis-ci.org/apache/arrow/jobs/465515849 - This appears to be a Travis / environment hiccup, and not a result of my change.  Could someone restart this build?

wesm commented:
You can trigger a fresh build by `git commit --amend` followed by `git push -f`. I'm restarting this build entry

wesm commented:
@siddharthteotia @atuldambalkar can you review this?

atuldambalkar commented on java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/JdbcToArrowUtils.java:
```diff
@@ -222,17 +259,38 @@ private static void allocateVectors(VectorSchemaRoot root, int size) {
+    Preconditions.checkNotNull(root, "Vector Schema cannot be null");
     Preconditions.checkNotNull(calendar, "Calendar object can't be null");
 
+    jdbcToArrowVectors(rs, root, new JdbcToArrowConfig(new RootAllocator(0), calendar));
```
I think, RootAllocator with argument 0, may not behave as expected. Have you tested this?

mikepigott replied:
This is a little confusing.  The config's allocator is used in the `JdbcToArrow` static member functions to construct a `VectorSchemaRoot`.  At this point, the `VectorSchemaRoot` has already been constructed, so the allocator is an unused argument.  That's why its allocation can be `0`.

atuldambalkar replied:
Yea, got it. Mainly  for the backeward compatibilit for JdbcToArrowUtils.java class.

emkornfield commented on java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/JdbcToArrow.java:
```diff
@@ -110,13 +112,36 @@ public static VectorSchemaRoot sqlToArrow(
+  /**
+   * For the given SQL query, execute and fetch the data from Relational DB and convert it to Arrow objects.
+   *
+   * @param connection Database connection to be used. This method will not close the passed connection object.
```
it might be a simpler contract to pass through the ResultSet object instead here?

mikepigott replied:
That version exists on line 224.

This version can certainly be removed, but I was trying to preserve the pair of contracts the original author supplied.  The two contracts were:

1. <Database + Query> + [Allocator | Calendar | Both | Neither]
2. <ResultSet> + [Allocator | Calendar | Both | Neither]

Since I'm adding a new contract of <ResultSet> + <Config>, it made sense to me to also add <Database + Query> + <Config>.

emkornfield commented on java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/JdbcToArrow.java:
```diff
@@ -89,7 +89,9 @@ public static VectorSchemaRoot sqlToArrow(Connection connection, String query, B
 
-    return sqlToArrow(connection, query, allocator, Calendar.getInstance(TimeZone.getTimeZone("UTC"), Locale.ROOT));
+    JdbcToArrowConfig config =
+            new JdbcToArrowConfig(allocator, Calendar.getInstance(TimeZone.getTimeZone("UTC"), Locale.ROOT), false);
```
Would a utility method for getting the UTC calendar make sense?

mikepigott replied:
Good call.  Done.

emkornfield commented on java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/JdbcToArrowTestHelper.java:
```diff
@@ -166,6 +173,41 @@ public static void assertNullValues(BaseValueVector vector, int rowCount) {
+      assertNotNull(metadata);
+      assertEquals(4, metadata.size());
+
+      assertEquals(rsmd.getCatalogName(i + 1), metadata.get(Constants.SQL_CATALOG_NAME_KEY));
```
might be cleaner if you start the loop at 1 instead of doing the addition in 4 places.

mikepigott replied:
Yup, I agree.  I had to use `i - 1` when getting the field metadata from the `VectorSchemaRoot`, but I agree that's 3 fewer lines of code with an offset.  Updated.

mikepigott commented:
Thanks for the code review!  I believe I've responded to all of the feedback.

praveenbingo commented on java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/JdbcToArrowConfig.java:
```diff
@@ -0,0 +1,149 @@
+   *
+   * @return Whether this configuration is valid.
+   */
+  public boolean isValid() {
```
older comment..please ignore..

praveenbingo commented on java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/JdbcToArrowConfig.java:
```diff
@@ -0,0 +1,149 @@
+   * @return This instance of the <code>JdbcToArrowConfig</code>, for chaining.
+   * @exception NullPointerExeption if <code>calendar</code> is <code>null</code>.
+   */
+  public JdbcToArrowConfig setCalendar(Calendar calendar) {
```
older comment..please ignore..

mikepigott commented on java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/JdbcToArrowUtils.java:
```diff
@@ -148,75 +159,93 @@ public static Schema jdbcToArrowSchema(ResultSetMetaData rsmd, JdbcToArrowConfig
+        metadata.put(Constants.SQL_TYPE_KEY, rsmd.getColumnTypeName(i));
+
+      } else {
+        metadata = null;
```
I can do that, but then I can't make the metadata variable final.  I've worked on teams in the past where the best practice was to mark things final as often as possible; is that the same with Arrow?

mikepigott commented on java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/JdbcToArrowUtils.java:
```diff
@@ -148,75 +159,93 @@ public static Schema jdbcToArrowSchema(ResultSetMetaData rsmd, JdbcToArrowConfig
+
+      final Map<String, String> metadata;
+      if (config.getIncludeMetadata()) {
+        metadata = new HashMap<String, String>();
```
Good point; will fix tonight.

mikepigott replied:
Fixed.

mikepigott commented on java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/JdbcToArrowConfig.java:
```diff
@@ -70,4 +73,13 @@ public Calendar getCalendar() {
+   *
+   * @return <code>true</code> to include field metadata, <code>false</code> to exclude it.
+   */
+  public boolean getIncludeMetadata() {
```
Good point; will fix tonight.

mikepigott replied:
Fixed.

mikepigott commented on java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/JdbcToArrowConfigBuilder.java:
```diff
@@ -98,6 +137,6 @@ public JdbcToArrowConfigBuilder setCalendar(Calendar calendar) {
    */
   public JdbcToArrowConfig build() {
-    return new JdbcToArrowConfig(allocator, calendar);
+    return new JdbcToArrowConfig(allocator, calendar, includeMetadata);
```
That makes sense. Will change tonight.

mikepigott replied:
Hang on, I'm confused - `includeMetadata` is initialized to `false` in the builder.  This is the behavior you asked for, right?  Only people who call `setIncludeMetadata(true)` will have the metadata generated.

praveenbingo replied:
Ah I saw the ctors using the false flag in other places and was misled..this looks ok.

mikepigott commented:
Thanks for the feedback!  I believe I've responded to all of it.

xhochy commented:
@mikepigott Can you rebase? Then I would merge this.

mikepigott commented:
Sure thing; I'll merge master when I get home tonight. Thanks!

mikepigott commented:
Merged.  Thanks for the review!

mikepigott commented:
I noticed in the merge that one part of #3066 was lost in the shuffle, and I just added it back in.  The calendar's time zone is used to create Timestamp types, even though it [is not needed](https://github.com/apache/arrow/blob/apache-arrow-0.11.1/java/vector/src/main/java/org/apache/arrow/vector/types/Types.java#L807-L822).  I'm not sure why it was lost, but I just added it back.

xhochy commented:
Thanks!
